### PR TITLE
Windows: Start of enabling tests under integration

### PR DESCRIPTION
**Pull request by salah-khan**
Integration tests on Windows.

This was originally @jhowardmsft's work and then vdemeesters and enables the integration tests on windows CI (see #36748)
This is also a follow up of #35362 — initial description below.
Currently in progress, with a custom job in jenkins.

This change is a requirement for #34226.
This change also brings the windowsRS1 build in line with all the other builds, which have an entry point at hack/ci. The windowsRS1 config has been updated to call this script.
This change is important because currently the configuration in jenkins makes too many assumptions about the repo and artifacts created by the build. We need a minimal interface between our CI configuration and the repo, so that we are free to make changes in a PR without requiring complicated acrobatics to support those changes in CI.

Signed-off-by: John Howard jhoward@microsoft.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like build is failing

File: integration/network/service_test.go [hunk 1]
```diff
@@ -314,7 +154,133 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
 	}
 }
 
+func TestDaemonRestartWithLiveRestore(t *testing.T) {
```
Comment (thaJeztah) [inline integration/network/service_test.go]: Why were these moved? Could we keep them at the same location to minimise the diff?

File: integration/plugin/logging/validation_test.go [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// +build !windows
```
Comment (thaJeztah) [inline integration/plugin/logging/validation_test.go]: Should we rename as well (`validation_linux_test.go` instead of a build-flag?

File: integration/container/stop_windows_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	"gotest.tools/assert"
+	"gotest.tools/poll"
+	"gotest.tools/skip"
+)
+
+// TestStopContainerWithTimeout checks that ContainerStop with
+// a timeout works as documented, i.e. in case of negative timeout
+// waiting is not limited (issue #35311).
+func TestStopContainerWithTimeout(t *testing.T) {
+	skip.If(t, testEnv.OSType == "windows")
```
Comment (thaJeztah) [inline integration/container/stop_windows_test.go]: This doesn't look right; this is in a `_windows` file, but skipped?

**Comment by jhowardmsft** [comment]
Ah, just looked briefly and I think hack/ci/windowsRS1.ps1 (which really should just be windows.ps1 as the same code has worked from RS1 through RS5, and continues to work on 19H1 builds internally). Further, it needs to be re-synced to the current master which you take either from the Jenkins RS1 configuration, or the source of truth at https://github.com/jhowardmsft/docker-w2wCIScripts/blob/master/runCI/executeCI.ps1

File: integration/container/stop_windows_test.go (hunk 3 above)
Comment (salah-khan) [reply integration/container/stop_windows_test.go]: This is being temporarily skipped because the timings are a problem. There are differences in Linux and Windows error codes and handling etc., however, some additional work is required to get the timing right.

File: integration/build/build_test.go [hunk 4]
```diff
@@ -166,16 +169,27 @@ func TestBuildMultiStageParentConfig(t *testing.T) {
 	resp.Body.Close()
 	assert.NilError(t, err)
 
+	time.Sleep(30 * time.Second)
```
Comment (thaJeztah) [inline integration/build/build_test.go]: Was this sleep needed on Linux as well?

File: integration/build/build_test.go [hunk 5]
```diff
@@ -282,14 +296,16 @@ func TestBuildWithEmptyLayers(t *testing.T) {
...
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows", "FIXME")
 	defer setupTest(t)()
 	// test both metadata and layer based commands as they may be implemented differently
 	dockerfile := `FROM busybox AS stage1
 ONBUILD RUN echo 'foo' >somefile
 ONBUILD ENV bar=baz
 
 FROM stage1
-RUN cat somefile # fails if ONBUILD RUN fails
+# fails if ONBUILD RUN fails
```
Comment (thaJeztah) [inline integration/build/build_test.go]: Was this change needed to make this work, or just a nit?

File: integration/network/service_test.go [hunk 6]
```diff
@@ -187,7 +150,58 @@ func TestDaemonWithBipAndDefaultNetworkPool(t *testing.T) {
 	delInterface(t, defaultNetworkBridge)
 }
 
+func TestServiceWithDefaultAddressPoolInit(t *testing.T) {
```
Comment (thaJeztah) [inline integration/network/service_test.go]: oh, can you move this one to its old position as well?

File: integration/internal/requirement/requirement_linux.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
+// +build !windows
```
Comment (thaJeztah) [inline integration/internal/requirement/requirement_linux.go]: This can be removed (file already as `_linux.go` suffix)

File: integration/internal/requirement/requirement_windows.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// +build windows
```
Comment (thaJeztah) [inline integration/internal/requirement/requirement_windows.go]: This can be removed (file already as `_windows.go` suffix)

File: integration/network/helpers_windows.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
+// +build windows
```
Comment (thaJeztah) [inline integration/network/helpers_windows.go]: This can be removed (file already as `_windows.go` suffix)

File: integration/network/macvlan/macvlan_test.go [hunk 10]
```diff
@@ -17,7 +19,7 @@ import (
 
 func TestDockerNetworkMacvlanPersistance(t *testing.T) {
 	// verify the driver automatically provisions the 802.1q link (dm-dummy0.60)
-	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
Comment (thaJeztah) [inline integration/network/macvlan/macvlan_test.go]: This can be removed now that we have the build-tag

File: integration/network/macvlan/macvlan_test.go [hunk 11]
```diff
@@ -42,7 +44,7 @@ func TestDockerNetworkMacvlanPersistance(t *testing.T) {
 }
 
 func TestDockerNetworkMacvlan(t *testing.T) {
-	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
Comment (thaJeztah) [inline integration/network/macvlan/macvlan_test.go]: This can be removed now that we have the build-tag

File: integration/network/service_test.go [hunk 12]
```diff
@@ -26,6 +26,7 @@ func delInterface(t *testing.T, ifName string) {
 }
 
 func TestDaemonRestartWithLiveRestore(t *testing.T) {
+	skip.If(t, testEnv.OSType == "windows")
```
Comment (thaJeztah) [inline integration/network/service_test.go]: Actually; looks like none of these tests are run on Windows, so perhaps these skips should be removed, and a build-tag (and/or rename the file to `_linux_test.go`).

OTOH, a `skip` will show in the test logs (giving an indication how many tests were skipped 🤔

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: integration/build/build_test.go (hunk 5 above)
Comment (salah-khan) [reply integration/build/build_test.go]: I'm not sure. This was a change made either by John Howard or Vincent. I don't know why. I simply picked up from where they left off.

File: integration/network/service_test.go (hunk 12 above)
Comment (salah-khan) [reply integration/network/service_test.go]: Ok, I'll leave them in place.

File: integration/network/service_test.go (hunk 1 above)
Comment (salah-khan) [reply integration/network/service_test.go]: Moved back.

File: integration/build/build_test.go (hunk 4 above)
Comment (salah-khan) [reply integration/build/build_test.go]: It was added by either Vincent or JHoward so I'm really not certain. Things work with it and presumably didn't work without it, but I'm not certain of the origin. It can be re-evaluated later it need be.

File: integration/plugin/logging/validation_test.go (hunk 2 above)
Comment (salah-khan) [reply integration/plugin/logging/validation_test.go]: I'm not certain. This code uses plug-ins which are not currently supported on Windows, however, if they are eventually supported, this may be possible to use directly at that time.

**Review by vdemeester — APPROVED** [review]
SGTM :tiger:

**Review by thaJeztah — APPROVED** [review]
SGTM as well

I'm not good at PowerShell, so perhaps @jhowardmsft can give this a final check, and merge if it looks good?

**Review by lowenna — CHANGES_REQUESTED** [review]
See https://github.com/moby/moby/pull/37715#issuecomment-424830351. I'd also need confirmation that Invoke-DockerCI hasn't broken as this has the potentially to break all internal testing if it does. Certainly need that sync to be done before this can be merged. Also one verification from @johnstep

**Comment by jhowardmsft** [comment]
cc @psftw too. This has knock on effect on CE/EE jobs too on the other Jenkins system.

File: integration/internal/requirement/requirement_linux.go (hunk 7 above)
Comment (salah-khan) [reply integration/internal/requirement/requirement_linux.go]: Done.

**Comment by jhowardmsft** [comment]
OK, have verified it doesn't break `Invoke-Docker-CI`. However, note that *none of the integration tests are actually run with this PR*, just the integration-cli tests. I would sign-off once the rename is done (and I'll need to update the Jenkins job to match). There's certainly some tidying to do in the CI script as put in this PR (arguably, it doesn't actually need to be in this PR as it's completely unrelated), as well as an update to Invoke-DockerCI to use the checked in script, and also removing the CI script from it's current master source of truth under my GH repo.

**Comment by ddebroy** [comment]
@jhowardmsft Thanks for verifying. I have done the renaming as requested.

**Comment by jhowardmsft** [comment]
Have updated the Jenkins job at https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/configure (you may not have access) to point to windows.ps1 and restarted the job to re-verify again, and ensure the console output contains `NEW entrypoint`.

Specifically

Old:
```
if (Test-Path hack\ci\windowsRS1.ps1) {
    Write-Host "NEW entrypoint"
    .\hack\ci\windowsRS1.ps1
    exit $LastExitCode
}
```

New:
```
if (Test-Path hack\ci\windows.ps1) {
    Write-Host "NEW entrypoint"
    .\hack\ci\windows.ps1
    exit $LastExitCode
}
```

Once this is merged, that `Write-Host` can be removed, along with the rest of the script embedded in the job. I'll also update the RS5 jobs.

@psftw - Have you managed to look at the follow-on impact to other CE/EE jobs?

**Comment by jhowardmsft** [comment]
LGTM. But I'm concerned about the impact to CE/EE jobs still which I don't have access to.

Have verified is running the in-repo script now after reconfiguring jenkins `Docker-PRs-WOW-RS1` job. From 
https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/22555/console:

```
00:00:58.114 [docker] $ powershell.exe -NonInteractive -ExecutionPolicy ByPass "& 'C:\Users\jenkins\jenkins2734567844840291984.ps1'"
00:00:58.495 NEW entrypoint
00:00:58.594 
```

**Review by lowenna — APPROVED** [review]

**Comment by ddebroy** [comment]
I synced up with @jhowardmsft offline about his concern around internal CI jobs. The plan is to start syncing up the enhancements made for them (RS3/4/5) into windows.ps1 once the above PR gets merged.
